Sources/Build/Sanitizers.swift

```diff
@@ -0,0 +1,117 @@
+
+ Copyright (c) 2018 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+import Basic
+import Utility
+
+/// A set of enabled runtime sanitizers.
+public struct EnabledSanitizers {
+    public enum Sanitizer: String {
+        case address
+        case thread
+    }
+
+    /// A set of enabled sanitizers.
+    private let sanitizers: Set<Sanitizer>
+
+    /// OS identification
+    private let triple: Triple
+
+    /// Runtime verbosity level
+    private let verbose: Bool
+
+    public init(from list: [Sanitizer] = [], triple: Triple = Triple.hostTriple, verbose: Bool = false) {
+        self.sanitizers = Set(list)
+        self.triple = triple
+        self.verbose = verbose
+    }
+
+    /// Return an established short name for a sanitizer, e.g. "asan".
+    private func shortname(_ sanitizer: Sanitizer) -> String {
+        switch sanitizer {
+        case .address: return "asan"
+        case .thread: return "tsan"
+        }
+    }
+
+    /// Return a dictionary with the sanitizer-specific list of additional
+    /// environment variables that have to be injected during runtime.
+    public func addRuntimeEnvironment(baseEnvironment: [String: String], toolchain: Toolchain) -> [String: String] {
+      #if os(macOS)
+        return macosAddRuntimeEnvironment(baseEnvironment: baseEnvironment, toolchain: toolchain)
+      #else
+        return [:]
+      #endif
+    }
+
+    private func macosAddRuntimeEnvironment(baseEnvironment: [String: String], toolchain: Toolchain) -> [String: String] {
+        var env: [String: String] = [:]
+        var insertLibs: [String] = []
+
+        if let existingLibs = baseEnvironment["DYLD_INSERT_LIBRARIES"] {
+            if existingLibs.count != 0 {
+                insertLibs = [existingLibs]
+            }
+        }
+
+        for sanitizer in sanitizers {
+            let libPath = sanitizerLibrary(clangPath: toolchain.clangCompiler, for: sanitizer)
+            insertLibs.append(libPath.asString)
+        }
+
+        env["DYLD_INSERT_LIBRARIES"] = insertLibs.joined(separator: ":")
+
+        if verbose {
+            env["DYLD_PRINT_LIBRARIES"] = "1"
+        }
+
+        return env
+    }
+
+    /// Sanitization flags for the C family compiler (C/C++)
+    public func compileCFlags() -> [String] {
+        return sanitizers.map({ "-fsanitize=\($0.rawValue)" })
+    }
+
+    /// Sanitization flags for the Swift compiler are the same as for
+    /// the linker, plus a set of -DSANITIZE_XXX defines for
+    /// source code level `#if SANITIZE_XXX`.
+    /// The C compiler can use `#if __has_feature(xxx_sanitizer)`
+    public func compileSwiftFlags() -> [String] {
+        return linkSwiftFlags() + sanitizers.map({"-DSANITIZE_\($0.rawValue.uppercased())"})
```

Comment:
@aciidb0mb3r please check this out.

Reply:
We shouldn't need to do this, compiler already has something for doing this check.

Reply:
https://clang.llvm.org/docs/AddressSanitizer.html#conditional-compilation-with-has-feature-address-sanitizer

Reply:
@aciidb0mb3r this is what is written in the comment above. The problem is that __has* thing is not accessible from the Swift due to the absence of a full-featured preprocessor.

Reply:
Ah I see. I don't know if we should do this. @ddunbar what do you think?

Reply:
I think it’s a reasonable thing to be able to consitionalize on. Let’s just make sure it gets documented somewhere in our Usage guide?

Reply:
hmm, we will need to figure out how will this interact with Xcode generation.

Reply:
Maybe someone should ask Kuba if there is any plan to expose a Swift level conditional? Ideally it’d be nice if SwiftPM didn’t need to impose a convention

Reply:
Ok, this ended up being a tarpit. I'll just remove the option. Whoever wants could always go `swift test -Xswiftc -DSANITIZER --sanitize=thread` or some such. If the official conditional becomes available in Swift, the users could use that directly.